client/src/components/ApplicationForm.tsx

```diff
@@ -1,110 +1,138 @@
-import { FormEvent, useContext, useState } from 'react';
-import { Job, submitApplication } from '../lib';
+import { FormEvent, useContext } from 'react';
+import { Job } from '../lib';
 import Button from './Button';
 import AppContext from './AppContext';
+import { Link } from 'react-router-dom';
 
 type ApplicationFormProps = {
   job: Job;
+  formState: Record<string, string>;
+  onInputChange: (key: string, value: string) => void;
+  onSubmit: (event: FormEvent<HTMLFormElement>) => void;
+  formSubmitted: boolean;
 };
 
-export default function ApplicationForm({ job }: ApplicationFormProps) {
-  const { user } = useContext(AppContext);
-  const [submitted, setSubmitted] = useState(false);
+export default function ApplicationForm({
+  job,
+  formState,
+  onInputChange,
+  onSubmit,
+  formSubmitted,
+}: ApplicationFormProps) {
+  type InputData = {
+    label: string;
+    type: string;
+  };
+
+  const formInputs: Record<string, InputData> = {
+    name: {
+      label: 'Name *',
+      type: 'text',
+    },
+    email: {
+      label: 'Email *',
+      type: 'email',
+    },
+    phone: {
+      label: 'Phone Number *',
+      type: 'number',
+    },
+    resume: {
+      label: 'Resume *',
+      type: 'file',
+    },
+    portfolioUrl: {
+      label: 'Portfolio URL *',
+      type: 'text',
+    },
+    githubUrl: {
+      label: 'Github URL *',
+      type: 'text',
+    },
+    proud: {
+      label:
+        'What is a project or piece of work you are most proud of and why? *',
+      type: 'textarea',
+    },
+    interesting: {
+      label:
+        "What is the most interesting thing about you that isn't on your resumé? *",
+      type: 'textarea',
+    },
+  };
 
+  const { user } = useContext(AppContext);
   const inputClasses =
     'rounded-lg p-2 bg-transparent border-[1.5px] border-stone-500 w-full mt-1';
 
-  async function handleSubmit(event: FormEvent<HTMLFormElement>) {
-    try {
-      event.preventDefault();
-      const formData = new FormData(event.currentTarget);
-      await submitApplication(formData);
-      setSubmitted(true);
-    } catch (err) {
-      console.error(err);
+  // Creating the form from the object above
+  const inputs = [];
+  for (let key in formInputs) {
```

Comment:
This is interesting and given that you are building the form yourself, it's certainly reasonable. It does, though, just move the long code from down in the `return JSX` to up here in the for-loop. So it depends on what part of the code you want to look nicest!

Large forms like this are often built with a form library because they are messy to create and manage. My favorite is React Hook Form. At some point you may want to look into that and see if it simplifies things for you.